### PR TITLE
Improved plugin support 3.0 

### PR DESCRIPTION
**Pull request by rjmatthews62**
Modified timesig help

Implemented TimeSig wrapper so random.qml would work.

Modified timesig help

* Improved support for MuseScore 3.0 plugins
* Added a "set" to ElementW for setting properties.
* Added a buildWrapper static function to allow different wrapper descendants.
* Converted "random.qml" to work with MuseScore 3.0
* Added a TimeSigW class to support added features required to manipulate TimeSig
* Added support for lists of valid property names for use with "get" and "set"

colornotes.qml now works with Musescore 3.0

* Improved support for MuseScore 3.0 plugins
* Added a "set" to ElementW for setting properties.
* Added a buildWrapper static function to allow different wrapper descendants.
* Converted "random.qml" to work with MuseScore 3.0
* Added a TimeSigW class to support added features required to manipulate TimeSig
* Added support for lists of valid property names for use with "get" and "set"
* Modified timesig help
colornotes.qml and random.qml now works with Musescore 3.0
Added proplist.qml to demonstrate readProperties and writeProperties

Formatting and Indenting tidy up.

Sorted out cursor.element()  call, fixed elementWrapper linking.

Fixed some merging problems.

### CODE REVIEW

File: libmscore/chord.cpp [hunk 1]
```diff
@@ -3400,6 +3400,29 @@ Shape Chord::shape() const
             });
       shape.add(ChordRest::shape());      // add articulation + lyrics
       return shape;
+}
```
Comment (Jojo-Schmitz) [inline libmscore/chord.cpp]: Wrong indent for this one curly brace

File: libmscore/chordrest.cpp [hunk 2]
```diff
@@ -1107,6 +1107,16 @@ bool ChordRest::setProperty(P_ID propertyId, const QVariant& v)
             }
       triggerLayout();
       return true;
+}
```
Comment (Jojo-Schmitz) [inline libmscore/chordrest.cpp]: Wrong indent for this one curly brace

File: libmscore/chordrest.cpp [hunk 3]
```diff
@@ -1517,6 +1527,11 @@ void ChordRest::removeMarkings(bool /* keepTremolo */)
       qDeleteAll(el());
       qDeleteAll(articulations());
       qDeleteAll(lyrics());
+}
```
Comment (Jojo-Schmitz) [inline libmscore/chordrest.cpp]: Wrong indent for this one curly brace

File: libmscore/cursor.cpp [hunk 4]
```diff
@@ -163,6 +217,10 @@ bool Cursor::nextMeasure()
 //   add
 //---------------------------------------------------------
 
+void Cursor::add(ElementW* s) {
```
Comment (Jojo-Schmitz) [inline libmscore/cursor.cpp]: I think that curly brace should be on a line of its oen (and be indented)

File: libmscore/duration.cpp [hunk 5]
```diff
@@ -177,6 +177,32 @@ bool DurationElement::setProperty(P_ID propertyId, const QVariant& v)
                   return Element::setProperty(propertyId, v);
             }
       return true;
+}
```
Comment (Jojo-Schmitz) [inline libmscore/duration.cpp]: Wrong indent for this one curly brace

File: libmscore/element.cpp [hunk 6]
```diff
@@ -1115,6 +1115,13 @@ bool Element::setProperty(P_ID propertyId, const QVariant& v)
       triggerLayout();
       setGenerated(false);
       return true;
+}
```
Comment (Jojo-Schmitz) [inline libmscore/element.cpp]: Wrong indent for this one curly brace

File: libmscore/note.cpp [hunk 7]
```diff
@@ -2540,6 +2540,15 @@ bool Note::setProperty(P_ID propertyId, const QVariant& v)
             }
       triggerLayout();
       return true;
+}
```
Comment (Jojo-Schmitz) [inline libmscore/note.cpp]: Wrong indent for this one curly brace

File: libmscore/score.cpp [hunk 8]
```diff
@@ -1637,6 +1637,12 @@ Segment* Score::lastSegment() const
       {
       Measure* m = lastMeasure();
       return m ? m->last() : 0;
+}
```
Comment (Jojo-Schmitz) [inline libmscore/score.cpp]: Wrong indent for this one curly brace

File: libmscore/scoreElement.cpp [hunk 9]
```diff
@@ -237,6 +242,16 @@ void ScoreElement::writeProperty(XmlWriter& xml, P_ID id) const
             }
       else
             xml.tag(id, getProperty(id), propertyDefault(id));
+}
```
Comment (Jojo-Schmitz) [inline libmscore/scoreElement.cpp]: Wrong indent for this one curly brace

File: libmscore/segment.cpp [hunk 10]
```diff
@@ -39,6 +39,10 @@
 
 namespace Ms {
 
+Segment* SegmentW::segment() {
```
Comment (Jojo-Schmitz) [inline libmscore/segment.cpp]: I think that curly brace should be on a line of its own (and be indented)

File: libmscore/timesig.cpp [hunk 11]
```diff
@@ -23,6 +23,13 @@
 
 namespace Ms {
 
+void TimeSigW::setSig(int numerator, int denominator, int st) {
```
Comment (Jojo-Schmitz) [inline libmscore/timesig.cpp]: I think that curly brace should be on a line of its own (and be indented)

File: libmscore/timesig.cpp [hunk 12]
```diff
@@ -458,6 +465,12 @@ bool TimeSig::setProperty(P_ID propertyId, const QVariant& v)
       return true;
       }
 
+void TimeSig::supportedProperties(QList<P_ID>& dest, bool writeable) {
```
Comment (Jojo-Schmitz) [inline libmscore/timesig.cpp]: same here

File: share/plugins/colornotes.qml [hunk 13]
```diff
@@ -12,29 +12,21 @@
...
-               "#009c95", // G
-               "#0071bb", // G#/Ab
-               "#5e50a1", // A
-               "#8d5ba6", // A#/Bb
-               "#cf3e96"  // B
+      property variant colors : [
+               "#e21c48", "#f26622", "#f99d1c",
+               "#ffcc33", "#fff32b", "#bcd85f",
+               "#62bc47", "#009c95", "#0071bb",
+               "#5e50a1", "#8d5ba6", "#cf3e96"
```
Comment (Jojo-Schmitz) [inline share/plugins/colornotes.qml]: Here you are still/again reverting a recent change (of mine)

Comment (Jojo-Schmitz) [reply share/plugins/colornotes.qml]: Here you are still/again reverting a recent change (of mine)
This is a serious issue, the rest is nitpicks

Comment (rjmatthews62) [reply share/plugins/colornotes.qml]: I swear to god it was the rebase messing with me. Most of these I've already fixed once.
Sigh. I'll look at them soon.

File: libmscore/chordrest.cpp [hunk 14]
```diff
@@ -1144,9 +1154,9 @@ bool ChordRest::isGrace() const
 bool ChordRest::isGraceBefore() const
       {
       return isChord()
-         && (toChord(this)->noteType() & (
-           NoteType::ACCIACCATURA | NoteType::APPOGGIATURA | NoteType::GRACE4 | NoteType::GRACE16 | NoteType::GRACE32
-           ));
+                  && (toChord(this)->noteType() & (
+                            NoteType::ACCIACCATURA | NoteType::APPOGGIATURA | NoteType::GRACE4 | NoteType::GRACE16 | NoteType::GRACE32
+                            ));
```
Comment (Jojo-Schmitz) [inline libmscore/chordrest.cpp]: The indent before looked OK to me...

File: libmscore/chordrest.cpp [hunk 15]
```diff
@@ -1156,7 +1166,7 @@ bool ChordRest::isGraceBefore() const
 bool ChordRest::isGraceAfter() const
       {
       return isChord()
-         && (toChord(this)->noteType() & (NoteType::GRACE8_AFTER | NoteType::GRACE16_AFTER | NoteType::GRACE32_AFTER));
+                  && (toChord(this)->noteType() & (NoteType::GRACE8_AFTER | NoteType::GRACE16_AFTER | NoteType::GRACE32_AFTER));
```
Comment (Jojo-Schmitz) [inline libmscore/chordrest.cpp]: same here

File: libmscore/timesig.h [hunk 16]
```diff
@@ -35,9 +36,26 @@ enum class TimeSigType : char {
 
 //---------------------------------------------------------------------------------------
 //   @@ TimeSig
-///    This class represents a time signature.
+//   This class represents a time signature.
```
Comment (Jojo-Schmitz) [inline libmscore/timesig.h]: I think the `///`is needed for the plugin documentation?

File: share/plugins/colornotes.qml [hunk 17]
```diff
@@ -74,18 +74,18 @@ MuseScore {
...
+                              if (cursor.element() && cursor.element().type == Ms.CHORD) {
+                                    var graceChords = cursor.element().graceNotes;
                                     for (var i = 0; i < graceChords.length; i++) {
                                           // iterate through all grace chords
-                                          var graceNotes = graceChords[i].notes;
-                                          for (var j = 0; j < graceNotes.length; j++)
-                                                func(graceNotes[j]);
+                                          var notes = graceChords[i].notes;
+                                          for (var j = 0; j < notes.length; j++)
+                                                func(notes[j]);
```
Comment (Jojo-Schmitz) [inline share/plugins/colornotes.qml]: this still is reverting a recent change of mine

File: share/plugins/colornotes.qml [hunk 18]
```diff
@@ -74,18 +74,18 @@ MuseScore {
                         if (fullScore)
                               cursor.rewind(0) // if no selection, beginning of score
 
-                        while (cursor.segment && (fullScore || cursor.tick < endTick)) {
-                              if (cursor.element && cursor.element.type === Element.CHORD) {
-                                    var graceChords = cursor.element.graceNotes;
+                        while (cursor.segment() && (fullScore || cursor.tick < endTick)) {
+                              if (cursor.element() && cursor.element().type == Ms.CHORD) {
```
Comment (Jojo-Schmitz) [inline share/plugins/colornotes.qml]: my recent change used a `===`

File: share/plugins/colornotes.qml [hunk 19]
```diff
@@ -74,18 +74,18 @@ MuseScore {
...
+                                          var notes = graceChords[i].notes;
+                                          for (var j = 0; j < notes.length; j++)
+                                                func(notes[j]);
                                     }
-                                    var notes = cursor.element.notes;
-                                    for (var k = 0; k < notes.length; k++) {
-                                          var note = notes[k];
+                                    var notes = cursor.element().notes;
+                                    for (var i = 0; i < notes.length; i++) {
+                                          var note = notes[i];
```
Comment (Jojo-Schmitz) [inline share/plugins/colornotes.qml]: this still is reverting a recent change of mine

File: share/plugins/colornotes.qml (hunk 17 above)
Comment (rjmatthews62) [reply share/plugins/colornotes.qml]: Yeah, that's on purpose. colornotes.qml has changed to support new syntax.

File: share/plugins/colornotes.qml (hunk 18 above)
Comment (rjmatthews62) [reply share/plugins/colornotes.qml]: That has changed on purpose... at least, the fact that the element types now have a different namespace, ie, Ms.CHORD and not Element.CHORD. (Not my idea, I'm building off previous work)
I don't see that the == vs === makes an important difference here.

File: share/plugins/colornotes.qml (hunk 19 above)
Comment (rjmatthews62) [reply share/plugins/colornotes.qml]: So above.

File: libmscore/chordrest.cpp (hunk 14 above)
Comment (rjmatthews62) [reply libmscore/chordrest.cpp]: Explain that the qtCreator Ctrl+I auto indenting.... (USING the preferred musescore settings file)

File: libmscore/timesig.h (hunk 16 above)
Comment (rjmatthews62) [reply libmscore/timesig.h]: That is a fair cop but society is to blame.

File: share/plugins/colornotes.qml (hunk 18 above)
Comment (Jojo-Schmitz) [reply share/plugins/colornotes.qml]: QtCreator hinted me at that it does make a (subtle) difference, that's why I changed it.

File: share/plugins/colornotes.qml (hunk 17 above)
Comment (Jojo-Schmitz) [reply share/plugins/colornotes.qml]: Not here. But I see you fixed it meanwhile, thanks ;-)

File: libmscore/timesig.h (hunk 16 above)
Comment (Jojo-Schmitz) [reply libmscore/timesig.h]: Sorry, I don't get that, is it an idiom that just doesn't translate for me?

Comment (rjmatthews62) [reply libmscore/timesig.h]: A Monty Python quote... :D

Comment (Jojo-Schmitz) [reply libmscore/timesig.h]: Ah, and poor me saw them dubbed in German ;-)

Comment (rjmatthews62) [reply libmscore/timesig.h]: I can see how it might be confusing in German... :P

**Comment by rjmatthews62** [comment]
Superseded by PR #3303

**Pull request closed without merging**